File: 131-palindrome-partitioning/131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> ds;
        
        func(0, s, ds, ans);
        
        return ans;
    }
    
    void func(int i, string s, vector<string> &ds, vector<vector<string>> &ans){
        if(i > s.size()-1){
            ans.push_back(ds);
            return;
        } 
            
        for(int j = i; j < s.size(); j++){
            
            if(isPal(s, i, j) == true){
                ds.push_back(s.substr(i, j -i + 1));
                func(j+1, s, ds, ans);
                ds.pop_back();
            }
        }
        
    }
    
    bool isPal(string s, int start, int end){
        
        while(start<=end){
            if(s[start++] != s[end--]) return false;
        }
        return true;
    }
};

File: 131-palindrome-partitioning/131-palindrome-partitioning.py
class Solution:
    def isPali(self, s):
        l,r = 0, len(s)-1
        while(l <= r):
            if s[l] != s[r]: return False
            l+=1
            r-=1
        return True
    
    
    def func(self, i, n, s, ds, ans):
        if i > n-1:
            ans.append(ds[:])
        
        for j in range(i, n):
            
            if(self.isPali(s[i:j+1])):
                sub = s[i:j+1]
                ds.append(sub)
                self.func(j+1, n, s, ds, ans)
                ds.pop()
    
    
    def partition(self, s: str) -> List[List[str]]:
        ds = []
        ans = []
        
        n = len(s)
        self.func(0, n, s, ds, ans)
        return ans